### ISSUE TITLE
Fatal error after upgrading to SMW 4.1.3 (ElasticSearch)

### ISSUE THREAD

**Issue by bertrandgorge** [SemanticMediaWiki/SemanticMediaWiki#5611]
### Setup

- SMW version: 4.1.3
- MW version:  1.39.6
- PHP version: 8.2.15
- DB system (MySQL, Blazegraph, etc.) and version: MySQL 5.7.44

### Issue
After upgrading to 4.1.3 (without any other changes), I get the following errors:

Upon runing update.php: 

```
...
...search_parameters table already exists.
[23-Feb-2024 10:19:48 UTC] PHP Deprecated:  Calling static trait method Elasticsearch\Namespaces\BooleanRequestWrapper::performRequest is deprecated, it should only be called on a class using the trait in /var/www/html/extensions/WikiSearch/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Namespaces/IndicesNamespace.php on line 218
Elasticsearch\Common\Exceptions\BadRequest400Exception from line 657 of /var/www/html/extensions/WikiSearch/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php: null
#0 /var/www/html/extensions/WikiSearch/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php(312): Elasticsearch\Connections\Connection->process4xxError(Array, Array, Array)
#1 /var/www/html/extensions/WikiSearch/vendor/react/promise/src/FulfilledPromise.php(28): Elasticsearch\Connections\Connection->Elasticsearch\Connections\{closure}(Array)
#2 /var/www/html/extensions/WikiSearch/vendor/ezimuel/ringphp/src/Future/CompletedFutureValue.php(65): React\Promise\FulfilledPromise->then(Object(Closure), NULL, NULL)
#3 /var/www/html/extensions/WikiSearch/vendor/ezimuel/ringphp/src/Core.php(341): GuzzleHttp\Ring\Future\CompletedFutureValue->then(Object(Closure), NULL, NULL)
#4 /var/www/html/extensions/WikiSearch/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php(235): GuzzleHttp\Ring\Core::proxy(Object(GuzzleHttp\Ring\Future\CompletedFutureArray), Object(Closure))
#5 /var/www/html/extensions/WikiSearch/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php(204): Elasticsearch\Connections\Connection->Elasticsearch\Connections\{closure}(Array, Object(Elasticsearch\Connections\Connection), Object(Elasticsearch\Transport), Array)
#6 /var/www/html/extensions/WikiSearch/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Transport.php(115): Elasticsearch\Connections\Connection->performRequest('HEAD', '/smw-data-wiki_...', Array, NULL, Array, Object(Elasticsearch\Transport))
#7 /var/www/html/extensions/WikiSearch/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Namespaces/BooleanRequestWrapper.php(44): Elasticsearch\Transport->performRequest('HEAD', '/smw-data-wiki_...', Array, NULL, Array)
#8 /var/www/html/extensions/WikiSearch/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Namespaces/IndicesNamespace.php(218): Elasticsearch\Namespaces\BooleanRequestWrapper::performRequest(Object(Elasticsearch\Endpoints\Indices\Exists), Object(Elasticsearch\Transport))
#9 /var/www/html/extensions/SemanticMediaWiki/src/Elastic/Connection/Client.php(748): Elasticsearch\Namespaces\IndicesNamespace->exists(Array)
#10 /var/www/html/extensions/SemanticMediaWiki/src/Elastic/Indexer/Rebuilder/Rollover.php(103): SMW\Elastic\Connection\Client->indexExists('smw-data-wiki_p...')
#11 /var/www/html/extensions/SemanticMediaWiki/src/Elastic/Installer.php(54): SMW\Elastic\Indexer\Rebuilder\Rollover->update('data')
#12 /var/www/html/extensions/SemanticMediaWiki/src/Elastic/ElasticStore.php(328): SMW\Elastic\Installer->setup()
#13 /var/www/html/extensions/SemanticMediaWiki/src/Store.php(490): SMW\Elastic\ElasticStore->setup(Object(SMW\Options))
#14 /var/www/html/includes/installer/DatabaseUpdater.php(547): SMW\Store::setupStore(true, Object(SMW\Options))
#15 /var/www/html/includes/installer/DatabaseUpdater.php(515): DatabaseUpdater->runUpdates(Array, true)
#16 /var/www/html/maintenance/update.php(209): DatabaseUpdater->doUpdates(Array)
#17 /var/www/html/maintenance/includes/MaintenanceRunner.php(309): UpdateMediaWiki->execute()
#18 /var/www/html/maintenance/doMaintenance.php(85): MediaWiki\Maintenance\MaintenanceRunner->run()
#19 /var/www/html/maintenance/update.php(319): require_once('/var/www/html/m...')
#20 {main}

```

When in Special:Version :

```
[6e8d31e4548b88411e23109a] /wiki/Sp%C3%A9cial:Version Elasticsearch\Common\Exceptions\BadRequest400Exception: 
{
  "error": {
    "root_cause": [
      {
        "type": "security_exception",
        "reason": "missing authentication token for REST request [/]",
        "header": {
          "WWW-Authenticate": "Basic realm=\"security\" charset=\"UTF-8\""
        }
      }
    ],
    "type": "security_exception",
    "reason": "missing authentication token for REST request [/]",
    "header": {
      "WWW-Authenticate": "Basic realm=\"security\" charset=\"UTF-8\""
    }
  },
  "status": 401
}

Backtrace:

from /var/www/html/extensions/WikiSearch/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php(657)
#0 /var/www/html/extensions/WikiSearch/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php(312): Elasticsearch\Connections\Connection->process4xxError(array, array, array)
#1 /var/www/html/extensions/WikiSearch/vendor/react/promise/src/FulfilledPromise.php(28): Elasticsearch\Connections\Connection->Elasticsearch\Connections\{closure}(array)
#2 /var/www/html/extensions/WikiSearch/vendor/ezimuel/ringphp/src/Future/CompletedFutureValue.php(65): React\Promise\FulfilledPromise->then(Closure, NULL, NULL)
#3 /var/www/html/extensions/WikiSearch/vendor/ezimuel/ringphp/src/Core.php(341): GuzzleHttp\Ring\Future\CompletedFutureValue->then(Closure, NULL, NULL)
#4 /var/www/html/extensions/WikiSearch/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php(235): GuzzleHttp\Ring\Core::proxy(GuzzleHttp\Ring\Future\CompletedFutureArray, Closure)
#5 /var/www/html/extensions/WikiSearch/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php(204): Elasticsearch\Connections\Connection->Elasticsearch\Connections\{closure}(array, Elasticsearch\Connections\Connection, Elasticsearch\Transport, array)
#6 /var/www/html/extensions/WikiSearch/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Transport.php(115): Elasticsearch\Connections\Connection->performRequest(string, string, array, NULL, array, Elasticsearch\Transport)
#7 /var/www/html/extensions/WikiSearch/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Client.php(1477): Elasticsearch\Transport->performRequest(string, string, array, NULL, array)
#8 /var/www/html/extensions/WikiSearch/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Client.php(756): Elasticsearch\Client->performRequest(Elasticsearch\Endpoints\Info)
#9 /var/www/html/extensions/SemanticMediaWiki/src/Elastic/Connection/Client.php(213): Elasticsearch\Client->info(array)
#10 /var/www/html/extensions/SemanticMediaWiki/src/Elastic/Connection/Client.php(854): SMW\Elastic\Connection\Client->info()
#11 /var/www/html/extensions/SemanticMediaWiki/src/Elastic/Connection/Client.php(195): SMW\Elastic\Connection\Client->isOpenSearch()
#12 /var/www/html/extensions/SemanticMediaWiki/src/MediaWiki/Hooks.php(1344): SMW\Elastic\Connection\Client->getSoftwareInfo()
#13 /var/www/html/includes/HookContainer/HookContainer.php(338): SMW\MediaWiki\Hooks->onSoftwareInfo(array)
#14 /var/www/html/includes/HookContainer/HookContainer.php(137): MediaWiki\HookContainer\HookContainer->callLegacyHook(string, array, array, array)
#15 /var/www/html/includes/HookContainer/HookRunner.php(3548): MediaWiki\HookContainer\HookContainer->run(string, array)
#16 /var/www/html/includes/specials/SpecialVersion.php(284): MediaWiki\HookContainer\HookRunner->onSoftwareInfo(array)
#17 /var/www/html/includes/specials/SpecialVersion.php(306): SpecialVersion::getSoftwareInformation()
#18 /var/www/html/includes/specials/SpecialVersion.php(183): SpecialVersion::softwareInformation()
#19 /var/www/html/includes/specialpage/SpecialPage.php(701): SpecialVersion->execute(NULL)
#20 /var/www/html/includes/specialpage/SpecialPageFactory.php(1428): SpecialPage->run(NULL)
#21 /var/www/html/includes/MediaWiki.php(316): MediaWiki\SpecialPage\SpecialPageFactory->executePath(string, RequestContext)
#22 /var/www/html/includes/MediaWiki.php(904): MediaWiki->performRequest()
#23 /var/www/html/includes/MediaWiki.php(562): MediaWiki->main()
#24 /var/www/html/index.php(50): MediaWiki->run()
#25 /var/www/html/index.php(46): wfIndexMain()
#26 {main}
```

```
```

Steps to reproduce the observation (recommendation is to use the [sandbox](https://sandbox.semantic-mediawiki.org)):

**Comment by bertrandgorge** [comment]
Looping @marijnvanwezel as there's some WikiSearch items in the call stack (even if the same WikiSearch version with SMW 4.1.2 works fine)

**Comment by JeroenDeDauw** [comment]
Did you upgrade MediaWiki or something else at the same time as upgrading SMW?

**Comment by jaideraf** [comment]
Also what is the ElasticSearch version?

**Comment by bertrandgorge** [comment]
Hello, and thanks for inquiring - here is the full version page https://wiki.tripleperformance.fr/wiki/Sp%C3%A9cial:Version (obviously I reverted from 4.1.3 to 4.1.2 so that it could work) - appart from the version of SMW everything else is as it is there.... No change on MW nor on anything else.

I've upgraded SMW using composer.

**Comment by jaideraf** [comment]
I do not know it it is related, but for recent MW, SMW and PHP versions, I would recommend you to use Elasticsearch 7.10.2, like Wikimedia/Wikipedia:
https://en.wikipedia.org/wiki/Special:Version

**Comment by bertrandgorge** [comment]
Actually I tried - because I didn't know if ElasticSearch would be installed with the right version by default, I installed it with 7 and when I saw that it was coming by default in 6.8.2 I reverted to that - but it didn't change anything... For what it's worth I also checked that elastic search would connect properly (if there was a doubt) and I can confirm that it's only the version of SMW that breaks it....

**Comment by marijnvanwezel** [comment]
How did you configure authentication to ElasticSearch? That has changed in the latest version of SMW:

```
$smwgElasticsearchCredentials = [
    'user' => 'username',
    'pass' => 'password',
];
```

See also #5451.

**Comment by marijnvanwezel** [comment]
The latest version of SMW requires ES 7.10.x by the way.

**Comment by jaideraf** [comment]
Also note the following warning "MediaWiki 1.39+ require Elasticsearch 7.10.2 (6.8.23+ is possible using a compatibility layer)" in the page about [CirrusSearch](https://www.mediawiki.org/wiki/Extension:CirrusSearch)

**Comment by jaideraf** [comment]
I confirm that Elasticsearch 7.10.2 is working fine with MW 1.39.6, SMW 4.1.3: https://wikincat.org/wiki/Special:Version

**Comment by bertrandgorge** [comment]
Ah... Thanks all of you. So I managed to get things working indeed, I just had to change the config to add the new configuration from Marjin, downgrade PHP from PHP8 to PHP7.4 and upgrade ElasticSearch from 6 to 7.10.2... The good news is that it works ! Note that I couldn't get things working without downgrading PHP.

Unless I'm mistaking, the change of configuration is not too much documented anywhere. I'll spend some time doing that in the next days (I'm in a train right now so my internet is not stable enough for editing the wiki)

Also, @marijnvanwezel it seems that WikiSearch is asking for ElasticSearch 6.x max (specifically "elasticsearch/elasticsearch": "^5.3|^6.0" in composer.json). I had to change that as well... (or should I point to another branch ? But es-810-mw-139 requires ES 8.10 so that's a bit too much too)....
